src/session.rs

```diff
@@ -607,45 +625,70 @@ where
         }
     }
 
-    pub(crate) async fn with_capabilities_and_connector(
+    pub(crate) async fn create_client_and_parse_url(
         webdriver: &str,
-        cap: &webdriver::capabilities::Capabilities,
         connector: C,
-    ) -> Result<Client, error::NewSessionError> {
+    ) -> Result<
+        (
+            hyper_util::client::legacy::Client<C, BoxBody<hyper::body::Bytes, Infallible>>,
+            url::Url,
+        ),
+        error::NewSessionError,
+    > {
         // Where is the WebDriver server?
-        let wdb = webdriver.parse::<url::Url>();
-        let wdb = wdb.map_err(error::NewSessionError::BadWebdriverUrl)?;
+        let wdb = webdriver
+            .parse::<url::Url>()
+            .map_err(error::NewSessionError::BadWebdriverUrl)?;
+
         // We want a tls-enabled client
         let client = hyper_util::client::legacy::Client::builder(TokioExecutor::new())
             .build::<_, BoxBody<hyper::body::Bytes, Infallible>>(connector);
 
-        let mut cap = cap.to_owned();
+        Ok((client, wdb))
+    }
+
+    pub(crate) async fn setup_session(
+        client: hyper_util::client::legacy::Client<C, BoxBody<hyper::body::Bytes, Infallible>>,
+        wdb: url::Url,
+        session_id: Option<&str>,
+        cap: Option<webdriver::capabilities::Capabilities>,
+    ) -> Result<Client, error::NewSessionError> {
         // We're going to need a channel for sending requests to the WebDriver host
         let (tx, rx) = mpsc::unbounded_channel();
 
-        // Set up our WebDriver session.
-        tokio::spawn(Session {
-            rx,
-            ongoing: Ongoing::None,
-            client,
-            wdb,
-            session: None,
-            is_legacy: false,
-            ua: None,
-            persist: false,
-        });
+        if let Some(session_id) = session_id {
+            // Reconnect to an existing session
+            tokio::spawn(Session::new(rx, client, wdb, session_id));
+        } else if let Some(_capabilities) = cap {
```

Comment:
Why guard this branch behind `cap.is_some()`?

Reply:
Got it, I've removed the check.